File: 300-longest-increasing-subsequence/longest-increasing-subsequence.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        List<Integer> piles = new ArrayList<>();
        for (int i=0; i<nums.length; i++) {
            int num = nums[i];
            int index = Collections.binarySearch(piles, num);
            if (index < 0) {
                index = -(index + 1);
            }
            if (index == piles.size()) {
                piles.add(num);
            } else {
                piles.set(index, num);
            }
        }
        return piles.size();
    }
}

File: 300-longest-increasing-subsequence/longest-increasing-subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [0]*n
        for i in reversed(range(n)):
            dp[i] = 1
            for j in range(i+1, n):
                if nums[j] > nums[i]:
                    dp[i] = max(dp[i], 1 + dp[j])
        return max(dp)